plasmapy/utils/decorators/checks.py

```diff
@@ -1,20 +1,1219 @@
+    checks_on_return
+        specified checks on the return of the wrapped function
+
+    **checks
+        specified checks on the input arguments of the wrapped function
+    """
+    def __init__(self, checks_on_return=None, **checks):
+        self._checks = checks
+        if checks_on_return is not None:
+            self._checks['checks_on_return'] = checks_on_return
+
+    @property
+    def checks(self):
+        """
+        Requested checks on the decorated function's input arguments
+        and/or return.
+        """
+        return self._checks
+
+
+class CheckValues(CheckBase):
+    """
+    A decorator class to 'check' -- limit/control -- the values of input and return
+    arguments to a function or method.
+
+    Parameters
+    ----------
+    checks_on_return: Dict[str, bool]
+        Specifications for value checks on the return of the function being wrapped.
+        (see `check values`_ for valid specifications)
+
+    **checks: Dict[str, Dict[str, bool]]
+        Specifications for value checks on the input arguments of the function
+        being wrapped.  Each keyword argument in `checks` is the name of a function
+        argument to be checked and the keyword value contains the value check
+        specifications.
+
+        .. _`check values`:
+
+        The value check specifications are defined within a dictionary containing
+        the keys defined below.  If the dictionary is empty or omitting keys,
+        then the default value will be assumed for the missing keys.
+
+        ================ ======= ================================================
+        Key              Type    Description
+        ================ ======= ================================================
+        can_be_negative  `bool`  [DEFAULT `True`] values can be negative
+        can_be_complex   `bool`  [DEFAULT `False`] values can be complex numbers
+        can_be_inf       `bool`  [DEFAULT `True`] values can be :data:`~numpy.inf`
+        can_be_nan       `bool`  [DEFAULT `True`] values can be :data:`~numpy.nan`
+        none_shall_pass  `bool`  [DEFAULT `False`] values can be a python `None`
+        ================ ======= ================================================
+
+    Notes
+    -----
+    * Checking of function arguments `*args` and `**kwargs` is not supported.
+
+    Examples
+    --------
+    .. code-block:: python
+
+        from plasmapy.utils.decorators.checks import CheckValues
+
+        @CheckValues(arg1={'can_be_negative': False, 'can_be_nan': False},
+                     arg2={'can_be_inf': False},
+                     checks_on_return={'none_shall_pass': True)
+        def foo(arg1, arg2):
+            return None
+
+        # on a method
+        class Foo:
+            arg1={'can_be_negative': False, 'can_be_nan': False},
+                         arg2={'can_be_inf': False},
+                         checks_on_return={'none_shall_pass': True)
+            def bar(self, arg1, arg2):
```

Comment:
Doesn't look like we're actually calling `@CheckValues` here!

Reply:
Bizarre! How did that happen? Ok, I’ll probably get to that fix tomorrow.